src/modular/boxed_residue.rs

```diff
@@ -21,6 +21,18 @@ use std::sync::Arc;
 #[cfg(feature = "zeroize")]
 use zeroize::Zeroize;
 
+fn div_by_2(a: &BoxedUint, modulus: &BoxedUint) -> BoxedUint {
+    let (half, is_odd) = a.shr1_with_carry();
+    let half_modulus = modulus.shr1();
+
+    let if_even = half.clone();
+    let if_odd = half
+        .wrapping_add(&half_modulus)
+        .wrapping_add(&BoxedUint::one_with_precision(a.bits_precision()));
+
+    BoxedUint::conditional_select(&if_even, &if_odd, is_odd)
```

Comment:
You can cut down on a number of allocations here:
```suggestion
    let (mut ret, is_odd) = a.shr1_with_carry();
    let half_modulus = modulus.shr1();

    let if_odd = ret
        .wrapping_add(&half_modulus)
        .wrapping_add(&BoxedUint::one());

    ret.conditional_assign(&if_odd, is_odd);
    ret
```